### PR TITLE
Reset state after #_digest

### PR DESCRIPTION
**Pull request by fanatid**

### CODE REVIEW

**Comment by dcousens** [comment]
Tests?

And what is the motivation?

**Comment by fanatid** [comment]
note: require major version bump

motivation: remove any data related with source

**Comment by dcousens** [comment]
@fanatid sounds good, but you should add a test to verify that :+1:

File: index.js [hunk 1]
```diff
@@ -6,6 +6,7 @@ function HashBase (blockSize) {
   Transform.call(this)
 
   this._block = new Buffer(blockSize)
+  this._block.fill(0)
```
Comment (dcousens) [inline index.js]: Why is this necessary?

File: index.js [hunk 2]
```diff
@@ -72,6 +73,12 @@ HashBase.prototype.digest = function (encoding) {
 
   var digest = this._digest()
   if (encoding !== undefined) digest = digest.toString(encoding)
+
+  // reset state
+  this._block.fill(0)
+  this._blockOffset = 0
+  for (var i = 0; i < 4; ++i) this._length[i] = 0
```
Comment (dcousens) [inline index.js]: Why not just `this.length = [0,0,0,0]`

**Comment by dcousens** [comment]
Also,  what are the performance implications?

**Comment by dcousens** [comment]
ping in @dominictarr

File: index.js (hunk 1 above)
Comment (fanatid) [inline index.js]: because it's makes internal state indistinguishable, internal state before and after using should be same

**Comment by fanatid** [comment]
of course performance is worse
and I not sure, how much really we need this, because I didn't see any crypto related implementation which do this
if we really need this we also should be fix other crypto-browserify libraries (in feature at least)

File: index.js (hunk 1 above)
Comment (dcousens) [inline index.js]: So the idea is to hide whether the `_block` has been used at all?

Comment (fanatid) [inline index.js]: over-write any potential private information -- you mean on buffer allocation?

**Comment by fanatid** [comment]
ping @calvinmetcalf @indutny 
What you think about reset internal state after work is completed? (i.e. remove any private information)
Not only here, also in other libraries like `elliptic` for example.

File: index.js (hunk 1 above)
Comment (calvinmetcalf) [inline index.js]: ok so this one is unnecessary since this wouldn't be private information we needed to hide.

hashes can come in 3 states

- not used
- in progress
- completed 

currently there is no way to distinguish the 3 by looking at the buffer, this pull would change it so that completed and not used were zero filled while in progress is not on the grounds that 

 > internal state before and after using should be same

I have no reason to believe that and while I'd be happy to be convinced otherwise it just seems like cargo culting some pattern or something

File: index.js [hunk 3]
```diff
@@ -72,6 +73,12 @@ HashBase.prototype.digest = function (encoding) {
 
   var digest = this._digest()
   if (encoding !== undefined) digest = digest.toString(encoding)
+
+  // reset state
+  this._block.fill(0)
```
Comment (calvinmetcalf) [inline index.js]: good catch here

File: index.js (hunk 1 above)
Comment (indutny) [inline index.js]: Buffer's come 0-filled now, not sure it is reasonable.

**Comment by fanatid** [comment]
Thank you all for your comments

**Comment by dcousens** [comment]
@fanatid I'd be happy with the clear by default,  but the `fill(0)` at the start is unnecessary,  and any way,  its default for `Buffer` to be zeroed now anyway! :smile:

**Comment by fanatid** [comment]
Yes, clear by default is unnecessary. But `Buffer` always zero filled only with flag `--zero-fill-buffers` otherwise `Buffer.allocUnsafe` can return not zero filled buffer.

**Comment by dcousens** [comment]
ACK, PR LGTM
